### ISSUE TITLE
The type of nullable<T> and vector<T> fields are indistinguishable using only a SchemaDef

### ISSUE THREAD

**Issue by csdahlberg** [microsoft/bond#161]
We work with some schemas which we have a SchemaDef for, but not the .bond file (or the corresponding C# class) used to originally create the SchemaDef. Any fields with a type of `nullable<T>` in the original .bond file show up as a `BT_LIST` in the SchemaDef.

For example, here is a struct defined in a .bond file:

```
struct NullableAndVector
{
    1: nullable<int64> nullableInt;
    2: vector<int64> vectorInt;
}
```

The properties of the resulting C# class match up perfectly:

```
[global::Bond.Id(1), global::Bond.Type(typeof(global::Bond.Tag.nullable<long>))]
public long? nullableInt { get; set; }

[global::Bond.Id(2)]
public List<long> vectorInt { get; set; }
```

However, for both `nullableInt` and `vectorInt`, the SchemaDef (originally obtained using `Schema<NullableAndVector>.RuntimeSchema.SchemaDef`) shows both fields as being a `BT_LIST` with `BT_INT64` elements, and all metadata is identical.

Is there any way to distinguish between `nullable<T>` and `vector<T>` using only information in a SchemaDef? If not, can a change be made to make that possible?

**Comment by chadwalters** [comment]
This equivalence is a bit of an artifact of a design decision (alluded to in [Issue 148](https://github.com/Microsoft/bond/issues/148)) that makes list<T>, vector<T>, and nullable<T> equivalent on the wire. Also, the SchemaDef basically contains the bare minimum to work with untagged protocols and bonded<> and no more; it is not an encapsulation of all potentially interesting metadata from the IDL.

Perhaps you could describe a bit more about your scenario? That would help us figure out an appropriate solution.

**Comment by csdahlberg** [comment]
We have web services that allow clients to save and retrieve client-generated Bond objects (in their Bond-serialized form). Something like:

1. Client creates `Foo.bond`
2. Client calls a web service to register the schema (passing in the Bond-serialized form of `Schema<Foo>.RuntimeSchema.SchemaDef`)
3. Client calls a web service to save an instance of `Foo` (passing in the Bond-serialized form of a `Foo` object and the ID of the `Foo` schema)
4. Server sometimes deserializes `Foo` to run some validations.

Step 4 is done using `RuntimeDeserializer` from [my fork](https://github.com/csdahlberg/bond). It takes a `RuntimeSchema` and either a `IClonableTaggedProtocolReader` or `IClonableUntaggedProtocolReader` as inputs, and effectively returns a dictionary containing the field names and values. It mostly works, but can make some things much more complicated to work with, since all nullable properties get deserialized as lists.

As a side note, I previously submitted those changes in pull request #130, which was (rightfully) declined. I do hope to eventually re-implement those changes in a manner consistent with existing code (i.e. expression trees), but the ambiguity between `nullable` and `vector` would still be an issue. Resolving that ambiguity seems like a prerequisite for any proper run-time-only C# implementation.

**Comment by chadwalters** [comment]
Cool use of Bond -- very interesting.

I am a bit hesitant to extend the SchemaDef to support this one bit of disambiguation. As I mentioned, the SchemaDef isn't designed to capture all potentially interesting metadata from the IDL. That said, if we can come up with a principled way to add more info that enables a large class of interesting functionality (i.e.: not just a spot fix for this particular issue) without bloating the data on the wire signicantly, I'd be open to exploring it.

**Comment by csdahlberg** [comment]
The only thing that's come to mind so far that _might_ fit your description is adding something like a `MaxLength` field to `TypeDef`. That would work for me, since I could treat `vector<T>` with `MaxLength` of 1 as `nullable<T>`. It could also potentially be used to restrict string length, blob size, etc.

**Comment by chadwalters** [comment]
Interesting suggestion. We are considering other approaches as well. I'll try to post some proposals next week.

**Comment by csdahlberg** [comment]
Anything new with this?

**Comment by chadwalters** [comment]
Chris, sorry for the long delay -- I know you are looking for a fix here. We've decided that the best approach is simply add an additional field to bond::TypeDef that will hold the original IDL type. The default value will be "unknown" for backwards compatibility. There will be a lot of redundancy with TypeDef.id field but it will provide for disambiguation of nullable/list/vector. Thoughts/comments?

**Comment by csdahlberg** [comment]
That sounds great.

I don't know what the plans are for implementation, but let me know if there's anything you'd like me to help with.

**Comment by lalo** [comment]
We are working on the implementation, just to make sure @csdahlberg, knowing the difference between (nullable) and (list or vector) is enough for you right? With the current solution we have we cannot distinguish between a list or a vector.

**Comment by csdahlberg** [comment]
Correct, we do not need to distinguish between `list<T>` and `vector<T>`. We handle them the same when we do not have a C# class for the Bond struct we are trying to deserialize.

**Comment by lalo** [comment]
@csdahlberg : This is our current proposal:

```
// Enumerator to distinguish the different subtypes treated as BT_LIST
enum ListSubType
{
    // Not a BT_LIST or info not available to infer the ListSubType
    UNKNOWN = 0,

    // IDL type is of nullable type
    NULLABLE = 1,

    // IDL type is of blob type
    BLOB = 2
}
```

and adding this to the TypeDef struct:

```
    // Distinguish between BondDataTypes that are markerd as BT_LIST; see ListSubType
    5: ListSubType list_sub_type;
```


Objects serialized with the old schema will always be marked as `UNKNOWN `and the recommendation is that you treat them as lists. Objects with the new change will be marked if nullable or blob type and if they are lists or vectors they will show as `UNKNOWN `and should be treated as lists. This will streamline the different code paths when checking for the `ListSubType`.

Consider how this affects migrating old schemas to new schemas and if this info is enough for you.

The other alternative that we are evaluating is adding a `LISTORVECTOR`.

```
// Enumerator to distinguish the different subtypes treated as `BT_LIST`
enum ListSubType
{
    // Not a BT_LIST or info not available to infer the ListSubType
    UNKNOWN = 0,

    // IDL type is of nullable type
    NULLABLE = 1,

    // IDL type is of blob type
    BLOB = 2,

    // IDL type is list or vector type
    LISTORVECTOR = 3
}
```

This way one can know for sure if the data they are dealing with is serialized with the old Bond (always will be marked as `UNKNOWN`) while in theory complicating the logic on how to deal with `UNKNOWN` and `LISTORVECTOR`.

We are leaning towards the first proposal but want to hear your thoughts about both of the options.

**Comment by csdahlberg** [comment]
The first proposal seems to be the better option to me, as well. `LISTORVECTOR` would be either redundant (if it is already known that all schemas have this new field populated) or not provide a complete solution (if there are any existing schemas that do not have this new field populated).

Depending on how long the implementation of this new field takes, migration to the new schemas might be trivial for us. Even if it takes a while, we can work in such a way that migration shouldn't be too bad.

**Comment by chwarr** [comment]
Bond version 4.3.0 is now live on NuGet.org. These changes are included in that version.

**Comment by chwarr** [comment]
We've gotten reports of backward compatibility problems with 4.3.0 due to the added field in `TypeDef`. We're backing out that change until we can address these compatibility issues.

**Comment by lalo** [comment]
Worth mentioning that we still have the functionality during runtime as outlined in this example in c# 
https://github.com/Microsoft/bond/blob/b6df00909702fd154066fd5842a9cdd89c190c6e/cs/test/internal/ReflectionTests.cs#L59

**Comment by ramon-garcia** [comment]
I am also having ambiguity between nullable and vector.

A solution that preserves full backward compatibility would be to add metadata to any field that happens to be a nullable.

**Comment by ramon-garcia** [comment]
Another possible solution that preserves full backward compatibility:

- Add a new enum to BondDataType BT_NULLABLE. No new field
- In order to avoid breaking existing code, either a flag to gbc, or even a new name for Nullable (such as Nullable2). 

With this method, old code will use the old serialization format. Only new protocols, when they choose to do so, will use the new serialization method for Nullable

**Comment by chwarr** [comment]
Our current plan is to re-add the `list_sub_type` field after the incompatible user has fixed and deployed their bug. This work is in progress.

**Comment by ramon-garcia** [comment]
Ok. Thank you.

**Comment by chwarr** [comment]
We believe that the incompatible user has deployed their fix broadly enough that these SchemaDef fields can be restored.

**Comment by chwarr** [comment]
Update from 2023: this is not a change that the Bond project plans to make now or in the foreseeable future. While we were able to fix some of the incompatible apps, we deemed changing runtime schemas too risky due to the unknown apps that could be broken by such a change. I'm going to close this issue as "Won't Fix", bot because it isn't a good idea, but because of compatibility concerns.

At this juncture, I would say it is best to consider the existing runtime schema schema (the runtime metaschema, if you will) frozen. There could be a `RuntimeSchema2` that the library would refuse to serialize in a non-tagged format at some time in the future, but Microsoft has no need for that and no plans to implement something like it.
